src/Type/ClosureType.php

```diff
@@ -334,7 +334,7 @@ public function inferTemplateTypes(Type $receivedType): TemplateTypeMap
 			return $receivedType->inferTemplateTypesOn($this);
 		}
 
-		if ($receivedType->isCallable()->no()) {
+		if ($receivedType->isCallable()->no() || ! $receivedType instanceof self) {
```

Comment:
Hi, what about the same change and test in CallableType.php? :)

Reply:
But I thought that can be a valid use case? `new Foo` which is `ObjectType`, can be callable, and we should try to infer template types. I did this for `ClosureType` because nothing else can be `Closure` other than `Closure` 😄 

Or did I understand this wrong?

Reply:
I added the test cases. But the same change did not work well in `CallableType`. Handled a bit differently and it works now.

Also added a test case with an invokable class, seems to work as expected 👍🏽